libCEED solvers

barker29 opened the pull request:
This PR uses solvers based on the libCEED element-basis decomposition to solve Poisson-like problems. The main addition is an `AlgebraicCeedSolver` class which is able to automatically generate a preconditioner for a high-order operator that uses the Ceed backend, with whatever integrators are available in the Ceed backend and with varying coefficients. This works in parallel and on (multiple) GPUs, though the coarsest solve on GPUs is suboptimal.

One of the main questions for discussion is how much of the new code in `fem/libceed` really belongs in MFEM and how much should be put elsewhere. In addition, any general comments on the ideas and structure of the interface are welcome.
<!--GHEX{"id":1763,"author":"barker29","editor":"tzanio","reviewers":["YohannDudouit","tzanio","v-dobrev","jandrej"],"assignment":"2020-10-28T19:49:04-07:00","approval":"2021-03-09T23:24:56.939Z","merge":"2021-03-21T22:37:42.764Z"}XEHG--><!--GHEXTABLE-->
 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1763](https://github.com/mfem/mfem/pull/1763) | @barker29 | @tzanio | @YohannDudouit + @tzanio + @v-dobrev + @jandrej | 10/28/20 | 03/09/21 | 03/21/21 | |
<!--ELBATXEHG-->

barker29 commented:
Ping @YohannDudouit @jeremylt ; I would be interested in any thoughts you have about this. It clearly still needs a lot of work but I think there's enough here for you to provide useful feedback.

barker29 commented:
With @pazner 's parallel implementation, I think this is ready for review. Apologies to the reviewers for the large amount of code. If people think some of that code should go in libCEED itself, I am open to that.

dylan-copeland commented:
This is my first time ever to look at libCEED. Should I just build https://github.com/CEED/libCEED and then build MFEM with `MFEM_USE_CEED=YES`? Does libCEED need special build options to enable GPU computing?

tzanio commented:
> This is my first time ever to look at libCEED. Should I just build https://github.com/CEED/libCEED and then build MFEM with MFEM_USE_CEED=YES?

Yes, just keep it in `../libCEED`. If you have CUDA or libxsmm or MAGMA installed, it will pick those up.

dylan-copeland commented on fem/libceed/ceed-assemble.cpp:
```diff
@@ -0,0 +1,310 @@
+{
+   *(void **)p = realloc(*(void **)p, n*unit);
+   if (n && unit && !*(void **)p)
+      // LCOV_EXCL_START
```
Remove?

barker29 replied:
Done.

jeremylt commented on fem/libceed/ceed-assemble.cpp:
```diff
@@ -0,0 +1,310 @@
+         {
+            case CEED_EVAL_NONE:
+            case CEED_EVAL_INTERP:
+               ierr = CeedHackRealloc(numemodein + 1, &emodein); CeedChk(ierr);
```
You can include `ceed-backend.h` or we can promote the alloc/realloc functions to `ceed.h` to avoid this

pazner replied:
I think the problem is that some of these functions are marked with hidden visibility.

jeremylt replied:
Ah yes. That still might be changeable if it's desirable.

barker29 commented:
On lassen, the command line `lrun -n1 -c1 -g1 ./ex1p -pa -d ceed-cuda -o 4 -a` is producing the following error:
```
CUDA error: (cudaMemcpy(dst, src, bytes, cudaMemcpyHostToDevice)) failed with error:
 --> an illegal memory access was encountered
 ... in function: void *mfem::CuMemcpyHtoD(void *, const void *, size_t)
 ... in file: general/cuda.cpp:86
```
I am pretty sure this command line worked at some point in the development of this branch. I will try to figure out what is going on, but I am very bad at debugging on GPU so if anybody has any feeling for what is happening I'd appreciate the thoughts.

pazner commented:
> On lassen, the command line `lrun -n1 -c1 -g1 ./ex1p -pa -d ceed-cuda -o 4 -a` is producing the following error:
> 
> ```
> CUDA error: (cudaMemcpy(dst, src, bytes, cudaMemcpyHostToDevice)) failed with error:
>  --> an illegal memory access was encountered
>  ... in function: void *mfem::CuMemcpyHtoD(void *, const void *, size_t)
>  ... in file: general/cuda.cpp:86
> ```
> 
> I am pretty sure this command line worked at some point in the development of this branch. I will try to figure out what is going on, but I am very bad at debugging on GPU so if anybody has any feeling for what is happening I'd appreciate the thoughts.

This seems to be working for me…

barker29 commented:
Turns out my issue was related to using an old libCEED - a fresh checkout clears it up.

YohannDudouit commented on fem/libceed/ceed-assemble.cpp:
```diff
@@ -0,0 +1,310 @@
+            CeedInt din = -1;
+            for (int ein = 0; ein < numemodein; ++ein)
+            {
+               if (emodein[ein] == CEED_EVAL_GRAD)
+               {
+                  din += 1;
+               }
+               if (emodein[ein] == CEED_EVAL_INTERP)
+               {
+                  Bmat(numemodein * q + ein, n) += interpin[q * elemsize + n];
+               }
+               else if (emodein[ein] == CEED_EVAL_GRAD)
+               {
+                  Bmat(numemodein * q + ein, n) += gradin[(din*nqpts+q) * elemsize + n];
+               }
```
```suggestion
               if (emodein[ein] == CEED_EVAL_INTERP)
               {
                  Bmat(numemodein * q + ein, n) += interpin[q * elemsize + n];
               }
               else if (emodein[ein] == CEED_EVAL_GRAD)
               {
                  din += 1;
                  Bmat(numemodein * q + ein, n) += gradin[(din*nqpts+q) * elemsize + n];
               }
```

barker29 replied:
Done.

YohannDudouit commented on fem/libceed/ceed-assemble.cpp:
```diff
@@ -0,0 +1,310 @@
+namespace mfem
+{
+
+int CeedSingleOperatorFullAssemble(CeedOperator op, SparseMatrix *out)
```
@jedbrown @jeremylt Don't you think it would be good to have this directly in libCEED?

YohannDudouit commented on fem/libceed/ceed-wrappers.cpp:
```diff
@@ -0,0 +1,241 @@
+{
+   int ierr = 0;
+
+   const CeedScalar *x_ptr;
+   CeedScalar *y_ptr;
+   CeedMemType mem;
+   CeedGetPreferredMemType(internal::ceed, &mem);
+   if (Device::Allows(Backend::CUDA) && mem == CEED_MEM_DEVICE)
+   {
+      x_ptr = x.Read();
+      y_ptr = y.ReadWrite();
+   }
+   else
+   {
+      x_ptr = x.HostRead();
+      y_ptr = y.HostReadWrite();
+      mem = CEED_MEM_HOST;
+   }
+
+   ierr += CeedVectorSetArray(u_, mem, CEED_USE_POINTER,
+                              const_cast<CeedScalar*>(x_ptr));
+   ierr += CeedVectorSetArray(v_, mem, CEED_USE_POINTER, y_ptr);
+
+   ierr += CeedOperatorApply(oper_, u_, v_, CEED_REQUEST_IMMEDIATE);
+
+   ierr += CeedVectorTakeArray(u_, mem, const_cast<CeedScalar**>(&x_ptr));
+   ierr += CeedVectorTakeArray(v_, mem, &y_ptr);
```
It would be nice to just use `CeedAddMult`

barker29 replied:
Fixed, but my solution is a little bit of a hack. Let me know if you have a better idea.

barker29 replied:
I made some more changes and think it's better.

YohannDudouit commented on fem/libceed/ceed-wrappers.cpp:
```diff
@@ -0,0 +1,241 @@
+{
+   int ierr = 0;
+
+   const CeedScalar *x_ptr;
+   CeedScalar *y_ptr;
+   CeedMemType mem;
+   CeedGetPreferredMemType(internal::ceed, &mem);
+   if ( Device::Allows(Backend::CUDA) && mem==CEED_MEM_DEVICE )
+   {
+      x_ptr = x.Read();
+      y_ptr = y.ReadWrite();
+   }
+   else
+   {
+      x_ptr = x.HostRead();
+      y_ptr = y.HostReadWrite();
+      mem = CEED_MEM_HOST;
+   }
+
+   ierr += CeedVectorSetArray(u_, mem, CEED_USE_POINTER,
+                              const_cast<CeedScalar*>(x_ptr));
+   ierr += CeedVectorSetArray(v_, mem, CEED_USE_POINTER, y_ptr);
+
+   ierr += CeedInterpolationInterpolate(ceed_interp_, u_, v_);
+
+   ierr += CeedVectorTakeArray(u_, mem, const_cast<CeedScalar**>(&x_ptr));
+   ierr += CeedVectorTakeArray(v_, mem, &y_ptr);
```
It would be nice here also to use `CeedAddMult`

barker29 replied:
Fixed, but my solution is a little bit of a hack. Let me know if you have a better idea.

barker29 replied:
The pointwise mult of the vector makes this one a little harder to fix than the other. I think you're right that we should do something like this to avoid duplicating code, but I'll have to think about exactly how to restructure.

barker29 replied:
I didn't use `CeedAddMult` but I think I've improved the organization / modularity.

YohannDudouit commented on fem/libceed/ceed-wrappers.cpp:
```diff
@@ -0,0 +1,241 @@
+{
+   int ierr = 0;
+
+   const CeedScalar *x_ptr;
+   CeedScalar *y_ptr;
+   CeedMemType mem;
+   CeedGetPreferredMemType(internal::ceed, &mem);
+   if ( Device::Allows(Backend::CUDA) && mem==CEED_MEM_DEVICE )
+   {
+      x_ptr = x.Read();
+      y_ptr = y.ReadWrite();
+   }
+   else
+   {
+      x_ptr = x.HostRead();
+      y_ptr = y.HostReadWrite();
+      mem = CEED_MEM_HOST;
+   }
+
+   ierr += CeedVectorSetArray(v_, mem, CEED_USE_POINTER,
+                              const_cast<CeedScalar*>(x_ptr));
+   ierr += CeedVectorSetArray(u_, mem, CEED_USE_POINTER, y_ptr);
+
+   ierr += CeedInterpolationRestrict(ceed_interp_, v_, u_);
+
+   ierr += CeedVectorTakeArray(v_, mem, const_cast<CeedScalar**>(&x_ptr));
+   ierr += CeedVectorTakeArray(u_, mem, &y_ptr);
```
same.

YohannDudouit commented on fem/libceed/ceed.cpp:
```diff
@@ -692,6 +692,7 @@ void InitCeedBasisAndRestriction(const FiniteElementSpace &fes,
    {
       if (UsesTensorBasis(fes))
       {
+         const IntegrationRule &ir = IntRules.Get(Geometry::SEGMENT, irm.GetOrder());
```
I don't think you need this.

YohannDudouit commented on fem/libceed/ceed.hpp:
```diff
@@ -181,6 +181,20 @@ struct CeedMFOperator
 void InitCeedCoeff(Coefficient* Q, Mesh &mesh, const IntegrationRule &ir,
                    CeedData* ptr);
 
+#ifdef MFEM_USE_CEED
+/// Initialize a CeedBasis and a CeedElemRestriction
+void InitCeedBasisAndRestriction(const FiniteElementSpace &fes,
+                                 const IntegrationRule &ir,
+                                 Ceed ceed, CeedBasis *basis,
+                                 CeedElemRestriction *restr);
+
+void InitCeedTensorRestriction(const FiniteElementSpace &fes,
+                               Ceed ceed, CeedElemRestriction *restr);
+#endif
+
+/// Return the path to the libCEED q-function headers.
+const std::string &GetCeedPath();
+
```
Remove, this is already defined lower in the file.

YohannDudouit replied:
You can move `void InitCeedTensorRestriction(const FiniteElementSpace &fes, Ceed ceed, CeedElemRestriction *restr);`, lower in the file.

barker29 replied:
Done.

YohannDudouit commented on fem/libceed/ceedsolvers-atpmg.h:
```diff
@@ -0,0 +1,91 @@
+    Assumes a Gauss-Lobatto basis and tensor product elements.
+
+    This is a setup routine that operates on the host. */
+int CeedATPMGElemRestriction(int order,
```
These functions only use libCEED code, would it make sense to see if it should go in libCEED?

YohannDudouit commented on fem/libceed/ceedsolvers-utility.cpp:
```diff
@@ -0,0 +1,198 @@
+      return CeedError(ceed, 1, "Vector sizes don't match");
+   }
+
+   CeedMemType mem;
+   CeedGetPreferredMemType(ceed, &mem);
+   CeedScalar *a_data;
+   const CeedScalar *b_data;
+   ierr = CeedVectorGetArray(a, mem, &a_data); CeedChk(ierr);
+   ierr = CeedVectorGetArrayRead(b, mem, &b_data); CeedChk(ierr);
+   MFEM_FORALL(i, length,
+   {a_data[i] *= b_data[i];});
```
I don't think this is correct, I think you want to use the memory type that corresponds to MFEM (not libCEED, the `ceed` object) since you use MFEM_FORALL. So if MFEM is in device mode, then you should use `CEED_MEM_DEVICE` otherwise `CEED_MEM_HOST` for `mem`.

barker29 replied:
This makes sense, thanks, now fixed.

YohannDudouit commented on fem/libceed/ceedsolvers-interpolation.cpp:
```diff
@@ -0,0 +1,163 @@
+   const CeedScalar* fine_data;
+   ierr = CeedVectorGetArray(fine_multiplicity_r, CEED_MEM_HOST,
+                             &fine_r_data); CeedChk(ierr);
+   ierr = CeedVectorGetArrayRead(c_fine_multiplicity, CEED_MEM_HOST,
+                                 &fine_data); CeedChk(ierr);
+   for (int i = 0; i < height; ++i)
+   {
+      fine_r_data[i] = 1.0 / fine_data[i];
+   }
```
Do this with `MFEM_FORALL`?

barker29 replied:
Done.

YohannDudouit commented on fem/libceed/ceedsolvers-interpolation.cpp:
```diff
@@ -0,0 +1,163 @@
+
+   const CeedScalar *multiplicitydata, *indata;
+   CeedScalar *workdata;
+   CeedMemType mem;
+   CeedGetPreferredMemType(interp->ceed, &mem);
+   ierr = CeedVectorGetArrayRead(in, mem, &indata); CeedChk(ierr);
+   ierr = CeedVectorGetArrayRead(interp->fine_multiplicity_r, mem,
+                                 &multiplicitydata); CeedChk(ierr);
+   ierr = CeedVectorGetArray(interp->fine_work, mem, &workdata); CeedChk(ierr);
+   MFEM_FORALL(i, length,
+   {workdata[i] = indata[i] * multiplicitydata[i];});
```
`mem` does not have the right value, you need to set `mem` according the mode of `MFEM_FORALL`.

barker29 replied:
Done.

YohannDudouit reviewed (commented):
Overall this looks good, I think we can improve a few points, but nothing major. The main big change that I could see is moving some of this code to libCEED.
@jedbrown @jeremylt Do you think some of this should be moved to libCEED:
fem/libceed/ceedsolvers-atpmg.h
fem/libceed/ceed-assemble.hpp
fem/libceed/ceedsolvers-utility.h

pazner commented:
I think that after #1845 is merged (it is currently in next), this PR will require some minor refactoring, since it uses the `Multigrid` interface. I'm happy to work on this if you want, @barker29.

pazner commented:
> I think that after #1845 is merged (it is currently in next), this PR will require some minor refactoring, since it uses the `Multigrid` interface. I'm happy to work on this if you want, @barker29.

The only change that is really necessary to make this compatible with #1845 is to inherit from `GeometricMultigrid` instead of `Multigrid`, so the changes really are minimal. A more involved change would be to inherit from `Multigrid` directly, and change the algebraic spaces so they do not inherit from `FiniteElementSpace` (since they are not "full-fledged" finite element spaces, the inheritance isn't really "proper"), but does not need to happen now.

dylan-copeland commented on examples/ex1p.cpp:
```diff
@@ -204,7 +208,9 @@ int main(int argc, char *argv[])
    if (pa) { a.SetAssemblyLevel(AssemblyLevel::PARTIAL); }
+
    a.AddDomainIntegrator(new DiffusionIntegrator(one));
+   // a.AddDomainIntegrator(new MassIntegrator(one));
```
It looks like this is here for testing but should be removed before merging.

YohannDudouit commented on fem/libceed/ceed-wrappers.hpp:
```diff
@@ -0,0 +1,77 @@
+};
+
+/// Wraps CeedInterpolation object in an mfem::Operator
+class MFEMCeedInterpolation : public mfem::Operator
+{
+public:
+   MFEMCeedInterpolation(
+      Ceed ceed, CeedBasis basisctof,
+      CeedElemRestriction erestrictu_coarse,
+      CeedElemRestriction erestrictu_fine);
```
Would it be possible to fuse this class with `CeedInterpolation`?

barker29 replied:
Done.

barker29 commented on fem/libceed/ceed-wrappers.cpp:
```diff
@@ -0,0 +1,221 @@
+   MFEM_ASSERT(ierr == 0, "CEED error");
+}
+
+UnconstrainedMFEMCeedOperator::~UnconstrainedMFEMCeedOperator()
```
This class expects `oper_` to be constructed outside itself and passed in the constructor. I think whoever creates `oper_` should destroy it, and that's not this class.

barker29 commented on fem/ceed_algebraic.hpp:
```diff
@@ -0,0 +1,169 @@
+   }
+   ~AlgebraicSpaceHierarchy()
+   {
+      CeedElemRestrictionDestroy(&fine_er);
```
I can't really use the existing `InitCeedBasis` type functions, because they all assume the `FiniteElementSpace` has a `Mesh` but our algebraic spaces do not have a mesh. We could manually insert the basis/restrictions we build in `ceed_basis_map` and (maybe?) use the existing `RemoveCeedBasisAndRestriction` to delete them, but I see that as kind of a hack. I am open to cleaner ways of handling these pointers but we are interfacing with C code so I think some amount of manual creation/deletion is inevitable.

YohannDudouit replied:
As far as I know, the original purpose of these maps was to store the `CeedBasis` and `CeedRestriction` to avoid duplicating them in the case of multigrid, which seems to be exactly this case... From what I see, the `Mesh` is only used to access the dimension of the `FiniteElementSpace`, how do you access the dimension if you don't have a `Mesh`?

YohannDudouit replied:
@jeremylt Am I mistaken? My impression was that the purpose of your `CeedBasis` and `CeedElemRestriction` maps was targeting this specific case.

jeremylt replied:
Yeah, the idea behind the maps is to prevent duplication of `CeedBasis` and `CeedElemRestriction`s. Especially the non-strided restrictions, because the indices array on the fine grid be a very large amount of data and extra copies can fill up the GPU.

YohannDudouit replied:
I got the confirmation from @jeremylt that the `CeedElemRestriction` map was specifically designed to prevent duplicating the fine grid restrictions.

barker29 replied:
What exactly are you asking me to do? Write my own version of `InitCeedBasisAndRestriction` that does not require a full `FiniteElementSpace`, or rewrite the existing version, or manually insert the basis and restriction we use in the map?

barker29 replied:
I may have misunderstood you. You are only concerned about the finest `ElemRestriction`? For that one I can possibly use the existing routines.

YohannDudouit replied:
You were mentioning that these functions need a `Mesh` to access the dimension which is true, but if we could access the dimension of the `FiniteElementSpace` that would be sufficient and these functions wouldn't need a `Mesh`. Unfortunately, there doesn't seem to be a way to access the dimension of a `FiniteElementSpace` without the `Mesh`, wouldn't it make sense to create an empty `Mesh` just to have the dimension and the space dimension defined?

barker29 commented on fem/libceed/ceed-assemble.cpp:
```diff
@@ -0,0 +1,319 @@
+#ifdef MFEM_USE_CEED
+#include "ceedsolvers-utility.h"
+
+int CeedHackReallocArray(size_t n, size_t unit, void *p)
+{
+   *(void **)p = realloc(*(void **)p, n*unit);
+   if (n && unit && !*(void **)p)
+      return CeedError(NULL, 1, "realloc failed to allocate %zd members of size "
+                       "%zd\n", n, unit);
+   return 0;
+}
+
+#define CeedHackRealloc(n, p) CeedHackReallocArray((n), sizeof(**(p)), p)
+
+int CeedHackFree(void *p)
+{
+   free(*(void **)p);
+   *(void **)p = NULL;
+   return 0;
+}
```
I want to use `CeedRealloc`, but it is marked with hidden visibility. I am very open to other solutions here, I do not very much like copy/pasting this code from Ceed here. See also (https://github.com/mfem/mfem/pull/1763#discussion_r522518340)

jedbrown replied:
Why do you want `CeedRealloc`? We made it internal because we didn't want libCEED memory management idioms to leak out (or to encourage callers to adopt them). I'd think either MFEM has/can have a realloc idiom or it wouldn't be needed in this interface. Of course it's a different matter if the intent is to develop and test this code here and then submit to libCEED.

barker29 replied:
The code that uses `Realloc` is precisely the code (explicit assembly of a `CeedOperator`, used on the coarsest level so we can use conventional AMG) that we want to eventually see in libCEED itself. One option is to use a more MFEM-y idiom here and then change it again when submitting to libCEED, another is to try to push it in libCEED before we merge this in MFEM, but I am lazy and the code works so I am inclined to leave it more or less as is.

jedbrown replied:
In that case, I think it's fine to do what you've done since this code won't live in MFEM for long.

barker29 commented on fem/libceed/ceed-wrappers.hpp:
```diff
@@ -0,0 +1,77 @@
+namespace mfem
+{
+
+/// Wraps a CeedOperator in an mfem::Operator
+class MFEMCeedOperator : public Operator
```
Do you want me to do some of this here before merging or wait for #1924 ?

YohannDudouit replied:
To clarify a bit, I was thinking that this would be renamed `ConstrainedCeedOperator` and inherited from a `ConstrainedOperator` and that the constructor would use an `UnconstrainedCeedOperator`.

barker29 commented on fem/bilininteg.hpp:
```diff
@@ -1986,6 +1986,10 @@ class DiffusionIntegrator: public BilinearFormIntegrator
+
+   void SetupPA(const FiniteElementSpace &fes);
+
+   CeedData * GetCeedData() { return ceedDataPtr; }
```
I moved it to `BilinearFormIntegrator`.

tzanio commented:
@barker29 and @YohannDudouit -- please agree on the relation of this PR to https://github.com/mfem/mfem/pull/1766 and reflece that in the description.

YohannDudouit commented on fem/libceed/full-assembly.hpp:
```diff
@@ -0,0 +1,31 @@
+namespace mfem
+{
+
+/// Assembles (single processor) a CeedOperator as an mfem::SparseMatrix
```
```suggestion
/// Assembles (single processor) a CeedOperator as a mfem::SparseMatrix
```

tzanio replied:
as discussed on Slack, `an` is correct here.

YohannDudouit replied:
I would have thought that the article would have applied to the object not to the scope.

tzanio commented:
There is an ongoing work on this PR, so I am extending the approval due date to next week.

barker29 commented on fem/libceed/ceed-wrappers.cpp:
```diff
@@ -0,0 +1,322 @@
+   CeedMemType mem;
+};
+
+class UnconstrainedMFEMCeedOperator : public Operator
```
I tried this, it ended up pretty ugly but maybe you have a better way. The issue I have is that I need a separate `Mult()` methods available so I can wrap it in `ConstrainedOperator`, so the `ConstrainedMFEMCeedOperator` needs to own (not inherit from) some unconstrained operator.

YohannDudouit replied:
I'm a bit confused by what you are saying, isn't this `UnconstrainedMFEMCeedOperator` just an `MFEMCeedOperator` with `GetCeedOperator`, `Operator::SetupRAP`, and a new constructor?
```c++
class UnconstrainedMFEMCeedOperator : public MFEMCeedOperator
{
public:
   UnconstrainedMFEMCeedOperator(CeedOperator oper): MFEMCeedOperator(), oper(oper) { }
   CeedOperator GetCeedOperator() const { return oper; }
   using Operator::SetupRAP;
};
```

YohannDudouit commented on fem/libceed/operator.hpp:
```diff
@@ -0,0 +1,137 @@
+
+/** Wraps a CeedOperator in an mfem::Operator, with essential boundary
+    conditions. */
+class ConstrainedMFEMCeedOperator : public Operator
```
Shouldn't this class inherit from `ConstrainedOperator`?

barker29 replied:
Maybe, but the `ConstrainedOperator` constructor needs an `mfem::Operator` as an argument, so if I want to construct that MFEM operator here from a given `CeedOperator` I can't use inheritance. I am open to other designs but I don't think what you propose is a simple change.

YohannDudouit replied:
An `MFEMCeedOperator` is an `mfem::Operator`, wouldn't that work?

barker29 replied:
I don't understand. Are you proposing rewriting the `ConstrainedMFEMCeedOperator` constructor to take an `mfem::Operator` instead of a `CeedOperator` argument? I want the user to be able to just call this once and not construct a whole chain of objects.

YohannDudouit replied:
I'm just worried that we might want to use a `ConstrainedMFEMCeedOperator` *as a* `ConstrainedOperator`. If you tell me that this will never be the case because this will never be used outside the libCEED integration then I'm fine with it.

YohannDudouit commented on fem/libceed/solvers-utility.h:
```diff
@@ -0,0 +1,32 @@
+// Copyright (c) 2010-2020, Lawrence Livermore National Security, LLC. Produced
```
Can you rename this file to `util.hpp`, I don't think these functions are specific to solvers.

YohannDudouit commented on fem/libceed/operator.hpp:
```diff
@@ -0,0 +1,137 @@
+   void GetDiagonal(Vector &diag) const;
+   virtual ~MFEMCeedOperator()
+   {
+      if (owns_oper) { CeedOperatorDestroy(&oper); }
```
It might make sense that this class doesn't destroy `oper` by default, and that the class that inherit from it eventually do (I'm not a big fan of ownership flags).

YohannDudouit commented on fem/libceed/operator.hpp:
```diff
@@ -0,0 +1,137 @@
+   CeedOperator GetCeedOperator() const { return oper; }
+};
+
+class UnconstrainedMFEMCeedOperator : public MFEMCeedOperator
```
Should we remove the `MFEM` in `UnconstrainedMFEMCeedOperator` and `MFEMCeedInterpolation`, I added `MFEM` to `MFEMCeedOperator` because it conflicts with the `CeedOperator` defined by libCEED. My favorite option would have been to use a `ceed` namespace below the `mfem` namespace and rename all these objects to `ceed::Operator`, `ceed::UnconstrainedOperator`, `ceed::Interpolator`, and `ceed::VectorContext`...

barker29 replied:
Do you want me to add a `ceed` namespace in this PR or will you do it in #1766 ?

YohannDudouit replied:
I will do it in #1766 , I think it's simpler.

YohannDudouit reviewed (commented):
@tzanio @v-dobrev Would it be possible to introduce a `ceed` or `libceed` namespace? I think the code would be more readable and cleaner.

tzanio commented:
I'm OK with adding a `ceed` namespace inside the `mfem` namespace

barker29 commented:
@pazner somehow my merge of master into this branch in 394cdb2c messed up some of your changes in `multigrid.?pp` and `ex26`. I think I got it cleaned up, but could you take a quick look and make sure?

YohannDudouit commented on fem/ceed_algebraic.hpp:
```diff
@@ -0,0 +1,169 @@
+// Copyright (c) 2010-2020, Lawrence Livermore National Security, LLC. Produced
```
Is there a specific reason for this file to be in the `fem` directory instead of `fem/libceed`?

barker29 replied:
Originally the stuff in `fem/libceed` was more C-style, libceed-style stuff that I was considering contributing to libCEED but this file was always going to be an MFEM interface. All of that has changed a lot and I am open to moving this.

YohannDudouit replied:
Okay, I think it might indeed be cleaner to move it under `fem/libceed` in this case.

YohannDudouit commented on fem/ceed_algebraic.cpp:
```diff
@@ -0,0 +1,621 @@
+   for (int i = 0; i < num_integrators; ++i)
+   {
+      BilinearFormIntegrator *integ = (*bffis)[i];
+      count += TryToAddCeedSubOperator<DiffusionIntegrator>(integ, op);
+      count += TryToAddCeedSubOperator<MassIntegrator>(integ, op);
+      count += TryToAddCeedSubOperator<VectorDiffusionIntegrator>(integ, op);
+      count += TryToAddCeedSubOperator<VectorMassIntegrator>(integ, op);
+   }
```
Should all libCEED supported `Integrator` be tested here?

barker29 replied:
Yes, I think so, what am I missing? Convection is not merged yet, that's the only one I know of.

YohannDudouit replied:
It just has me thinking, it's bug prone (or omission prone), and it's silent if it fails, I would think that a virtual method in `NonlinearIntegrator` might make more sense, was this choice made to avoid polluting the `Integrator`s' code?

YohannDudouit commented on fem/libceed/operator.hpp:
```diff
@@ -0,0 +1,135 @@
+private:
+   Array<int> ess_tdofs;
+   const Operator *P;
+   class UnconstrainedMFEMCeedOperator *unconstrained_op;
```
Is the `class` keyword really needed?

YohannDudouit commented on fem/libceed/solvers-atpmg.cpp:
```diff
@@ -0,0 +1,728 @@
+   ierr = CeedElemRestrictionGetNumComponents(er_in, &numcomp); CeedChk(ierr);
+   if (numcomp != 1)
+   {
+      // todo: this will require more thought
+      return CeedError(ceed, 1, "Not implemented!");
```
Can you provide a more meaningful error message?

YohannDudouit commented on fem/libceed/solvers-atpmg.cpp:
```diff
@@ -0,0 +1,728 @@
+   }
+   ierr = CeedVectorRestoreArray(in_lvec, &lvec_data); CeedChk(ierr);
+
+   // todo: I am making assumptions about the ordering of the evec that in
+   // principle are decided by the backend, which I do not control
```
Could this result in silent bugs?

barker29 replied:
Yes, probably. It seems to work for all the backends I have tested.

jeremylt replied:
Evec ordering possibly being backend dependant is a core assumption in libCEED's internals

barker29 replied:
Is it entirely arbitrary at the global evector level? Can I at least assume that each element's dofs are placed together in the evector?

jeremylt replied:
https://libceed.readthedocs.io/en/latest/libCEEDdev/#internal-layouts

jeremylt replied:
Let me know if I can clarify anything in that - it's buried deep in our docs and not a lot of people who aren't me have read it

barker29 replied:
That's already very helpful, and will allow me to (a) check for the correct layout in some cases and (b) at least document what we're expecting so it's clear what's going on.

barker29 replied:
Clarification/Question: `CeedElemRestrictionCreate` takes an `offsets` array and documents its layout, which is independent of the backend. In practice I have found this layout is usually the same as the evector layout, but I know from the above this is not guaranteed. I guess I want to verify that `CeedElemRestrictionCreate` always takes its input in the same layout regardless of backend (ie, the backend might reorder it) and that the backend should call `CeedElemRestrictionSetELayout` to tell me how/if it reordered things.

jeremylt replied:
Yes, the `offsets` array always has the same structure.
```
  CeedInt layout[3] = {1, elemsize*nelem, elemsize};
  ierr = CeedElemRestrictionSetELayout(r, layout); CeedChk(ierr);
```
The vanilla CUDA backend uses the above layout, while the all CPU backends use
```
  CeedInt layout[3] = {1, elemsize, elemsize*ncomp};
  ierr = CeedElemRestrictionSetELayout(r, layout); CeedChk(ierr);
```
In theory, a backend need not use an evector that can be described like this. We don't have any yet, but @YohannDudouit has discussed creating a backend that does not use an evector layout that can be described in this way.

barker29 replied:
If some future backend does not use this organization, does `CeedElemRestrictionGetELayout()` return a nonzero error code? If so I can handle this situation relatively gracefully (it won't work, but at least the user knows it won't work).

jeremylt replied:
We haven't decided how that would work yet. We are about to revamp the error system, so the idea of recoverable errors might be added.

Currently though, you'd just get the array `[0, 0, 0]` back for such a backed.

jeremylt replied:
woops, I misrememberd

https://github.com/CEED/libCEED/blob/954a603368e46bb36631709b22e7d1287756ce4e/interface/ceed-elemrestriction.c#L202

We actually return a non-zero error code when there has been no ELayout set, but depending upon the error handler in use, this can be an abort.

YohannDudouit commented on fem/libceed/solvers-atpmg.cpp:
```diff
@@ -0,0 +1,728 @@
+   if (debug)
+      printf("coarse_P1d = %d, numelem = %d, coarse_elemsize = %d\n",
+             coarse_P1d, numelem, coarse_elemsize);
+   CeedInt * out_elem_dof = (CeedInt*) calloc(coarse_elemsize * numelem,
+                                              sizeof(CeedInt));
```
Why not `new CeedInt[coarse_elemsize * numelem]`?

YohannDudouit commented on fem/libceed/solvers-atpmg.cpp:
```diff
@@ -0,0 +1,728 @@
+   int running_out_ldof_count = 0;
+   if (dim == 2)
+   {
+      for (int e = 0; e < numelem; ++e)
```
Do you have any idea if this loop could use `MFEM_FORALL`?

YohannDudouit replied:
`MFEM_FORALL` is not about parallelism, it's about using a device.

barker29 replied:
To clarify, threading (on-node parallelism) will not work here because each iteration of the loop has to run sequentially, or else the produced numbering will not be correct.

YohannDudouit reviewed (commented):
Having a more detailed look at the multigrid functions' code I have the impression that there is a lot of unchecked (uncheckable?) assumptions in the code, is it correct?

barker29 commented on fem/libceed/util.cpp:
```diff
@@ -0,0 +1,205 @@
+   return 0;
+}
+
+/// assumes a tensor-product operator with one active field
```
I don't know! Probably there can only be one active field but that's why the comment is there because I wasn't sure.

jeremylt replied:
The similar code in libCEED assumes only one active input basis/restriction (but possibly with multiple qfunction fields with different basis evaluation modes). We are only coarsening the active input (the vector passed into CeedOperatorApply) as if the vector only contains nodes from one type of element.

So the original code in libCEED does not currently work for mixed finite elements or composite operators. Adding support for composite operators with separate prolongation operators for each sub operator's active input would be easy to add but I have not done it yet. Adding support for mixed finite elements would take more work to support in our automatic multigrid setup interface.

barker29 commented on fem/libceed/solvers-atpmg.cpp:
```diff
@@ -0,0 +1,728 @@
+   }
+}
+
+int CeedATPMGElemRestriction(int order,
```
I agree the function is quite complicated, it took me a long time to write and I am still not very happy with it. I did try to improve the comments, let me know if there is more I should do or other things I should clarify.

jandrej replied:
I agree with @barker29 here. I'm not happy with this either. But I also know how finicky restriction operators can be.

barker29 commented:
I tried to address most of the review comments. Outstanding issues include:

- [ ] `ex1`, `ex1p`, should the new algebraic solver be demonstrated in a `examples/ceed` subdirectory? My view is no, the whole point is this is a one-line addition to `ex1p` so that's where I want to put it, but I could be persuaded.
- [ ] Use `basis_map` and `restr_map` to handle memory for what is currently handled in `fine_er`. This is still on my todo list.
- [ ] List of `TryToAddCeedSubOperator` calls, can this be handled differently? I am open to suggestions but I was trying to avoid impacting the library too much, and I don't see an easy solution even by changing `BilinearForm`.
- [ ] Possibility of silent bugs based on assumptions on input data. This is a very real issue, and partly I think it is the nature of the solver because it intentionally takes very little input from the user and tries to guess how to produce a good solver automatically - it might guess wrong. But in some cases this can be probably be improved, by checking the layout of the Ceed operator etc. I will think about this some more but I am open to specific suggestions.

Anything else I am missing?

YohannDudouit commented on fem/libceed/util.h:
```diff
@@ -0,0 +1,32 @@
+#include <ceed-backend.h>
+
+int CeedVectorPointwiseMult(CeedVector a, const CeedVector b);
+int CeedOperatorGetActiveField(CeedOperator oper, CeedOperatorField *field);
+int CeedOperatorGetOrder(CeedOperator oper, CeedInt * order);
+int CeedOperatorGetActiveElemRestriction(CeedOperator oper,
+                                         CeedElemRestriction* restr_out);
+int CeedOperatorGetSize(CeedOperator oper, CeedInt * size);
+int CeedOperatorGetActiveBasis(CeedOperator oper, CeedBasis *basis);
```
All these functions should become methods of `mfem::ceed::Operator` (currently `MFEMCeedOperator`).

barker29 replied:
Again, these only act on Ceed objects, and they are not restricted to MFEM's implementation. In principle they could be used for example on a `CeedOperator` that comes from Petsc via PR #1963

YohannDudouit commented on fem/libceed/algebraic.cpp:
```diff
@@ -0,0 +1,620 @@
+}
+
+template <typename INTEG>
+int TryToAddCeedSubOperator(BilinearFormIntegrator *integ_in, CeedOperator op)
+{
+   INTEG *integ = dynamic_cast<INTEG*>(integ_in);
+   if (integ != NULL)
+   {
+      CeedCompositeOperatorAddSub(op, integ->GetCeedData()->oper);
+      return 1;
+   }
+   return 0;
```
I think the easiest is to replace this function with two methods in `NonlinearFormIntegrator`: `SupportsCeed()` and `AddToCompositeOperator(...)` that throws an error by default, and to overload `SupportsCeed()` in the `Integrator`s that support libCEED:
```c++
virtual bool NonlinearFormIntegrator::SupportsCeed()
{
   return false;
}

bool MassIntegrator::SupportsCeed()
{
   return true; // Maybe DeviceCanUseCeed()
}

bool VectorMassIntegrator::SupportsCeed()
{
   return true;
}

bool ConvectionIntegrator::SupportsCeed()
{
   return true;
}

bool DiffusionIntegrator::SupportsCeed()
{
   return true;
}

bool VectorDiffusionIntegrator::SupportsCeed()
{
   return true;
}

bool VectorConvectionNLFIntegrator::SupportsCeed()
{
   return true;
}

void NonlinearFormIntegrator::AddToCompositeOperator(CeedOperator op) // We might want to use mfem::ceed::Operator
{
   if (SupportsCEED())
   {
      CeedCompositeOperatorAddSub(op, oper);
   }
   else
   {
      MFEM_ABORT("libCEED does not support this Integrator");
   }
}
```

YohannDudouit replied:
Actually, `AddToCompositeOperator(Integrator &integ, CeedOperator &op)` as a function and just adding `SupportsCeed()` to NonlinearFormIntegrator` might be enough.

barker29 replied:
This is a very nice simplification, and I have implemented it. I am worried we just move the maintenance burden somewhere else (we have to remember to add `SupportsCeed()` to all the integrators when we add Ceed support) but it certainly looks much cleaner.

barker29 commented on fem/libceed/solvers-atpmg.cpp:
```diff
@@ -84,8 +87,9 @@ int CeedATPMGElemRestriction(int order,
    {
-      // todo: this will require more thought
-      return CeedError(ceed, 1, "Not implemented!");
+      // todo: multi-component will require more thought
+      return CeedError(ceed, 1, "Algebraic element restriction not "
+                       "implemented for multiple components.");
```
So should I also remove all the `CeedChk(ierr);` checks?

YohannDudouit replied:
I think it's better to do the `CeedChk(ierr)`, I should probably have done it too.

jeremylt replied:
For what its worth, `CeedChk()` is in the `ceed-backend.h` header, so its more intended for internal use.

We are revamping the error codes though (https://github.com/CEED/libCEED/pull/691), so it might make sense to pull the libCEED numeric error codes into MFEM's error handling system.

barker29 replied:
I think if it makes sense to use `MFEM_ABORT` instead of `return CeedError`, it also makes sense to change the return type to `void` or an object (as @YohannDudouit asks in other parts of the PR). But if I don't return `int` then `CeedChk` no longer makes sense. So I guess the overall question is how the two error systems interact?

jedbrown replied:
How does MFEM normally propagate errors from external libraries?

barker29 replied:
The Petsc interface does this:
```c++
// Error handling
// Prints PETSc's stacktrace and then calls MFEM_ABORT
// We cannot use PETSc's CHKERRQ since it returns a PetscErrorCode
#define PCHKERRQ(obj,err) do {                                                   \
     if ((err))                                                                  \
     {                                                                           \
        PetscError(PetscObjectComm((PetscObject)(obj)),__LINE__,_MFEM_FUNC_NAME, \
                   __FILE__,(err),PETSC_ERROR_REPEAT,NULL);                      \
        MFEM_ABORT("Error in PETSc. See stacktrace above.");                     \
     }                                                                           \
  } while(0);
```
which seems pretty reasonable. I can certainly put together a Ceed-like variant of this.

barker29 commented on examples/ex1p.cpp:
```diff
@@ -225,13 +229,23 @@ int main(int argc, char *argv[])
       if (UsesTensorBasis(fespace))
       {
-         prec = new OperatorJacobiSmoother(a, ess_tdof_list);
+#ifdef MFEM_USE_CEED
```
Aside from making `AlgebraicCeedSolver` a dummy object when `MFEM_USE_CEED` is not defined (which seems very ugly), I am not sure how to do this.

YohannDudouit replied:
I think it is still cleaner to have a dummy object that returns an error when attempted to be created without libCEED, rather than having these `#ifdef`.

barker29 commented on fem/fem.hpp:
```diff
@@ -62,4 +62,8 @@
 
+#ifdef MFEM_USE_CEED
+#include "libceed/algebraic.hpp"
+#endif
```
Done.

barker29 commented on fem/libceed/operator.cpp:
```diff
@@ -0,0 +1,356 @@
+
+
+/** Manages memory for using mfem::Vector s for Ceed operations */
+class MFEMCeedVectorContext
```
Removed.

barker29 commented on fem/libceed/solvers-atpmg.cpp:
```diff
@@ -438,7 +450,8 @@ int CeedATPMGElemRestriction(int order,
    else
    {
-      return CeedError(ceed, 1, "CeedATPMGElemRestriction does not yet support this dimension.");
+      return CeedError(ceed, 1,
+                       "CeedATPMGElemRestriction does not yet support this dimension.");
```
My model for what I am doing is that there are two types of code in this PR:

1. MFEM code that uses C++ conventions, often returns `void`, uses `MFEM_ABORT`, etc. as usual. Where this code calls a Ceed function, it uses `PCeedChk`. Anything that touches an existing MFEM object is in this category.
2. Interface code that only acts on or uses Ceed datatypes and Ceed objects. This code returns `int`, and uses `CeedChk()` and `CeedError` for error handling. 
 
This is consistent with many of the helper codes in for example `linalg/hypre_parcsr.cpp` and `linalg/petsc.cpp`. Maybe this is the wrong idea, but in my mind many of these routines feel like and should act like Ceed functions, even if they live in the MFEM repo.

barker29 commented on fem/libceed/operator.cpp:
```diff
@@ -0,0 +1,356 @@
+   ierr = CeedVectorGetArrayRead(c_fine_multiplicity, CEED_MEM_HOST,
+                                 &fine_data); CeedChk(ierr);
+   MFEM_FORALL(i, height,
+   {fine_r_data[i] = 1.0 / fine_data[i];});
```
I made this so it runs on CPU all the time. Since this is a setup routine, it is not performance critical for the solver.

barker29 commented on fem/libceed/operator.cpp:
```diff
@@ -0,0 +1,356 @@
+void MFEMCeedInterpolation::Mult(const mfem::Vector& x, mfem::Vector& y) const
+{
+   int ierr = 0;
+   MFEMCeedVectorContext context(x, y, u_, v_);
```
`MFEMCeedVectorContext` has been removed.

barker29 commented on fem/libceed/util.cpp:
```diff
@@ -0,0 +1,205 @@
+
+/// a = a (pointwise*) b
+/// @todo: using MPI_FORALL in this Ceed-like function is ugly
+int CeedVectorPointwiseMult(CeedVector a, const CeedVector b)
```
I disagree. This only acts on Ceed objects and so I think it should use Ceed conventions. Also, `*` has many meanings in linear algebra, so I think it's better to be explicit.

jandrej commented on fem/libceed/full-assembly.cpp:
```diff
@@ -0,0 +1,319 @@
+#ifdef MFEM_USE_CEED
+#include "util.h"
+
+int CeedHackReallocArray(size_t n, size_t unit, void *p)
+{
+   *(void **)p = realloc(*(void **)p, n*unit);
+   if (n && unit && !*(void **)p)
+      return CeedError(NULL, 1, "realloc failed to allocate %zd members of size "
+                       "%zd\n", n, unit);
+   return 0;
+}
+
+#define CeedHackRealloc(n, p) CeedHackReallocArray((n), sizeof(**(p)), p)
+
+int CeedHackFree(void *p)
+{
+   free(*(void **)p);
+   *(void **)p = NULL;
+   return 0;
+}
```
Is the name already indicating that a reviewer should ask about this? :)

barker29 replied:
Yes, and we've already had some discussions here. The hope is to eventually contribute the full assembly code to the Ceed library, in which case it will use `CeedRealloc` and `CeedFree`. Those are inaccessible in the MFEM context. We discussed contributing to Ceed first, then doing this PR, but decided it made sense to get everything up and running in MFEM first.

jandrej commented on fem/libceed/full-assembly.cpp:
```diff
@@ -0,0 +1,319 @@
+namespace mfem
+{
+
+int CeedSingleOperatorFullAssemble(CeedOperator op, SparseMatrix *out)
```
This seems to be quite useful function for all sorts of applications. I'd prefer to have this extensively documented where needed, since it also makes hard assumptions on memory residence.

barker29 replied:
I've improved the documentation in the `.hpp` file. Note that it always performs assembly on the host, but doesn't assume that's where the memory lives - it works fine for operators that are on device.

jandrej commented on fem/libceed/operator.cpp:
```diff
@@ -0,0 +1,359 @@
+namespace mfem
+{
+
+void MFEMCeedOperator::Mult(const Vector &x, Vector &y) const
```
The naming should be revised in view of Yohanns PR introducing the `CEED::` namespace

jandrej reviewed (commented):
I left some in-line comments.

I strongly agree with keeping the ceed library conventions in utility functions.

Can you explain more generally what the concept of the "composite" and "sub operators" is?

barker29 commented on examples/ex1p.cpp:
```diff
@@ -225,13 +229,23 @@ int main(int argc, char *argv[])
       if (UsesTensorBasis(fespace))
       {
-         prec = new OperatorJacobiSmoother(a, ess_tdof_list);
+#ifdef MFEM_USE_CEED
+         if (DeviceCanUseCeed() && algebraic_ceed)
+         {
+            prec = new AlgebraicCeedSolver(a, ess_tdof_list);
+         }
+         else
+#endif
```
I will certainly do this if desired. But I want to be sure: doing this will mean copying a 300-line file with changes to 6 lines into a new example subdirectory. Is that what we want?

v-dobrev replied:
This is a bit tricky because `AlgebraicCeedSolver` is not just a wrapper for something libCEED already does, so it is different in that regard from e.g. SUNDIALS or PETSc examples. That said, I also feel that keeping `ex1`+`ex1p` simple is better.

Maybe instead of creating a separate example, we can add this as an option in the multigrid example, `ex26`+`ex26p`, demonstrating another solver in MFEM. I'm not sure how well it will fit in there though.

barker29 commented on fem/bilininteg.hpp:
```diff
@@ -32,7 +32,10 @@ class BilinearFormIntegrator : public NonlinearFormIntegrator
+      : NonlinearFormIntegrator(ir), ceedDataPtr(NULL) { }
+
+   // CEED extension
+   CeedData* ceedDataPtr;
```
Previously this was in all the individual integrators that supported Ceed (ie, `MassIntegrator`, `DiffusionIntegrator`, etc.), not in the extensions. In particular this is really an Integrator, and the extensions are for the `BilinearForm` itself. I am certainly open to a different design but it's not clear to me how to do it.

YohannDudouit replied:
The reason for this is that, in the current implementation, a `CeedOperator` is really an `Integrator` not a `BilinearForm`, since the merge of #1766 this data is actually in `NonlinearFormIntegrator`. I think it would be challenging to have only one `CeedOperator` for the whole `BilinearFormExtension`, but it might be possible using a `CompositeOperator`... I don't think this should be done in this PR though...

v-dobrev replied:
There's already this (recently added with #1766): https://github.com/mfem/mfem/blob/b0770915511fc63fda3120b13c88db48946e4302/fem/nonlininteg.hpp#L32-L33
that is inherited by the `BilinearFormIntegrator` from the `NonlinearFormIntegrator`, so this is a second libCEED-related data member. What type of data is `ceedDataPtr` supposed to hold?

barker29 replied:
`ceedDataPtr` is not added in this PR, it was there before. What you're seeing in the diff is this pointer being moved, but it's probably irrelevant because Yohann's #1766 restructured everything, changed this name, made it an object, and moved it again, so when master is merged in here everything will make sense.

v-dobrev replied:
Okay, I thought `master` was already merged here (I didn't check).

It is probably best to merge it before further reviews because things will need to change anyway.

barker29 commented on fem/libceed/algebraic.cpp:
```diff
@@ -0,0 +1,627 @@
+
+#ifdef MFEM_USE_MPI
+
+class CeedAMG : public Solver
```
Yes, when we merge master with Yohann's PR now included we will make use of the `ceed` namespace. (This may be a major project, we'll see...)

barker29 commented on fem/libceed/algebraic.cpp:
```diff
@@ -0,0 +1,627 @@
+class CeedAMG : public Solver
+{
+public:
+   CeedAMG(ConstrainedMFEMCeedOperator &oper, HypreParMatrix *P)
```
I certainly don't need it, it is here just for modularity/clarity. Note that it is only in the `.cpp` file and is not a user-facing class.

The idea is to separate ownership and creation/destruction of some objects so that IMO `AlgebraicCeedMultigrid` is simpler and easier to understand. Would you like it better if it was a `BuildAMGFromCeed()` function analogous to the `BuildSmootherFromCeed()` function above?

jandrej replied:
I understand the scope and purpose now.

barker29 commented on fem/libceed/algebraic.cpp:
```diff
@@ -0,0 +1,627 @@
+   ierr = CeedCompositeOperatorCreate(internal::ceed, &op); PCeedChk(ierr);
+
+   // Get the domain bilinear form integrators (DBFIs)
+   Array<BilinearFormIntegrator*> *bffis = form.GetDBFI();
```
Boundary forms (ie DG) are not currently supported in Ceed. I will document this better and add an error in this case.

barker29 commented on fem/libceed/algebraic.cpp:
```diff
@@ -0,0 +1,627 @@
+      CeedOperator subop_coarse;
+      ierr = CeedATPMGOperator(subop, order_reduction, er, &basis_coarse,
+                               &basis_c2f, &subop_coarse); PCeedChk(ierr);
+      // destructions below make sense because these objects are
```
Not sure what you are asking. I added the comment because the objects in question stay around and in fact are used later in the code, so I wanted to remind myself and future maintainers what is going on (ie, we just change the refcount so a later "real" destruction cleanly deletes the objects).

barker29 commented:
I merged master in, and the new functionality works, but the merged code is not quite ready for review. I'll let you know when it's fully cleaned up.

barker29 commented:
I've cleaned this up, and I think I've addressed most reviewer comments. I'm sure there's more work to do after the big merge, please let me know what you think.

jandrej commented on fem/bilininteg.hpp:
```diff
@@ -2044,8 +2037,6 @@ class DiffusionIntegrator: public BilinearFormIntegrator
 
-   void SetupPA(const FiniteElementSpace &fes);
-
    bool SupportsCeed() const { return true; }
```
`DeviceCanUseCeed()` doesn't know if the `BilinearFormIntegrator` works with ceed though.

barker29 replied:
I agree with @YohannDudouit , the base class implementation returns false so changing the derived classes to return `DeviceCanUseCeed()` means `SupportsCeed()` only returns true if *both* Ceed is supported by device and by the integrator. So I made this change.

jandrej commented on fem/ceed/full-assembly.cpp:
```diff
@@ -0,0 +1,324 @@
+
+   // enforce structurally symmetric for later elimination
+   const int skip_zeros = 0;
+   MFEM_ASSERT(numemodein == numemodeout, "My undestanding fails in this case.");
```
typo: unde**r**standing -> or a proper failure message :D

barker29 commented on examples/ex1.cpp:
```diff
@@ -216,7 +216,7 @@ int main(int argc, char *argv[])
-         AlgebraicCeedSolver M(a, ess_tdof_list);
+         ceed::AlgebraicCeedSolver M(a, ess_tdof_list);
          PCG(*A, M, B, X, 1, 400, 1e-12, 0.0);
       }
```
I did this, which I agree makes the examples much cleaner. The cost is it makes the `ceed::AlgebraicSolver` implementation a bit uglier. I am still open to make an `examples/ceed` directory but I prefer the current approach.

YohannDudouit commented on examples/ex1.cpp:
```diff
@@ -213,23 +219,20 @@ int main(int argc, char *argv[])
    else
    {
-#ifdef MFEM_USE_CEED
       if (DeviceCanUseCeed() && algebraic_ceed)
```
You can remove `DeviceCanUseCeed()` here too.

YohannDudouit commented on examples/ex1p.cpp:
```diff
@@ -229,13 +235,11 @@ int main(int argc, char *argv[])
       if (UsesTensorBasis(fespace))
       {
-#ifdef MFEM_USE_CEED
          if (DeviceCanUseCeed() && algebraic_ceed)
```
ditto.

YohannDudouit commented on examples/ex1.cpp:
```diff
@@ -101,6 +105,12 @@ int main(int argc, char *argv[])
    //    CUDA, OCCA, RAJA and OpenMP based on command line options.
    Device device(device_config);
    device.Print();
+   if (algebraic_ceed)
+   {
+      MFEM_VERIFY(DeviceCanUseCeed(),
+                  "--algebraic makes no sense without Ceed backend!");
+      MFEM_VERIFY(pa, "--algebraic only makes sense with partial assembly");
+   }
```
I wish this could be checked inside the `ceed::AlgebraicSolver`, to keep `ex1` as clean as possible.

barker29 replied:
I moved the checks into the constructor, which makes `ex1` look nicer. One thing I don't like is now in some cases you can use `--algebraic` on the command line and it will just not do anything (for example without `-pa`), which might be confusing to users. I'm open to suggestions.

YohannDudouit replied:
It's a good point, however it's already the case with `-d ceed-cpu` for instance, but I agree that it would be nice to detect it in some way. BTW, shouldn't the algebraic solver work on any `CeedOperator`, not only the `-pa` ones? Have you tried on the `-mf` ones?

YohannDudouit replied:
I went ahead and tried with `AssemblyLevel::NONE` (commented line 937-938 in algebraic.cpp), and it seems to work just fine (I get the same number of iterations as with `AssemblyLevel::PARTIAL), could you check and make the change if you agree?

barker29 replied:
This surprised me at first but I think you're right; whatever assembly level Ceed supports can also be used by the solver.

YohannDudouit commented on fem/ceed/algebraic.cpp:
```diff
@@ -0,0 +1,989 @@
+#ifdef MFEM_USE_CEED
+
+/** Construct a ceed::Operator from a CeedOperator */
+class WrapOperator : public ceed::Operator
```
I would rename this class to `OwnedOperator` since this is what it seems to do differently to `Operator`.

YohannDudouit replied:
Actually, can't you just fuse this class with `ceed::Operator`?

YohannDudouit replied:
This class should also be move with `ceed/operator.{hpp,cpp}`.

barker29 replied:
I fused into `ceed::Operator`

YohannDudouit commented on fem/ceed/algebraic.cpp:
```diff
@@ -0,0 +1,989 @@
+   /** This object takes ownership of ceed_op and will delete it */
+   WrapOperator(CeedOperator ceed_op)
+   {
+      oper = ceed_op;
+      CeedElemRestriction er;
+      CeedOperatorGetActiveElemRestriction(oper, &er);
+      int s;
+      CeedElemRestrictionGetLVectorSize(er, &s);
+      height = width = s;
+      CeedVectorCreate(internal::ceed, height, &v);
+      CeedVectorCreate(internal::ceed, width, &u);
```
I wonder if this constructor should rather be in `ceed::Operator`.

YohannDudouit commented on fem/ceed/algebraic.cpp:
```diff
@@ -0,0 +1,989 @@
+
+/** Wraps a CeedOperator in an mfem::Operator, with essential boundary
+    conditions and a prolongation operator for parallel application. */
+class ConstrainedMFEMCeedOperator : public mfem::Operator
```
Rename to `ConstrainedOperator`.

YohannDudouit replied:
This class should also be move with `ceed/operator.{hpp,cpp}`.

barker29 replied:
I renamed it, but I didn't move it since I'm not sure it should be user-facing. Right now it is just helper code for the algebraic solver in the .cpp file.

YohannDudouit replied:
What would not make it good enough to be user-facing?

barker29 replied:
How many users want to take a `ceed::Operator`, an `Array` of essential boundary dofs, and an `Operator` representing the parallel prolongation operator to make an `mfem::Operator`? I just don't see this as a useful general class.

YohannDudouit commented on fem/ceed/algebraic.cpp:
```diff
@@ -0,0 +1,989 @@
+
+#ifdef MFEM_USE_MPI
+
+class CeedAMG : public Solver
```
I would rename to `AMG` or `AlgebraicMultigridSolver`.

barker29 replied:
I renamed it `AssembledAMG` which I think describes its purpose.

YohannDudouit commented on fem/ceed/util.hpp:
```diff
@@ -64,6 +74,14 @@ void InitBasisAndRestriction(const mfem::FiniteElementSpace &fes,
+void InitTensorRestriction(const FiniteElementSpace &fes,
+                           Ceed ceed, CeedElemRestriction *restr);
+
+int CeedOperatorGetActiveField(CeedOperator oper, CeedOperatorField *field);
```
Rename to `GetOperatorActiveField`?

barker29 replied:
As we've discussed before in other places, I think this should stay with the C-style and Ceed-style naming conventions.

YohannDudouit replied:
Not completely convinced, but fine.

YohannDudouit commented on fem/ceed/util.hpp:
```diff
@@ -64,6 +74,14 @@ void InitBasisAndRestriction(const mfem::FiniteElementSpace &fes,
+
+int CeedOperatorGetActiveField(CeedOperator oper, CeedOperatorField *field);
+
+int CeedOperatorGetActiveElemRestriction(CeedOperator oper,
```
Rename to `GetOperatorActiveElemRestriction`?

barker29 replied:
ditto.

barker29 commented on fem/ceed/util.hpp:
```diff
@@ -16,6 +16,7 @@
 #ifdef MFEM_USE_CEED
 #include <ceed.h>
 #include <ceed-hash.h>
+#include <ceed-backend.h>  // for CeedOperatorField (?)
```
The typedef `CeedOperatorField` isn't defined unless you include this, and I had a hard time isolating all the uses of this to the .cpp files (that typedef is used in the interfaces of several of these utility functions). I think you are right that it shouldn't be here in the .hpp file but I can't figure out how to fix it.

barker29 commented on fem/ceed/algebraic.hpp:
```diff
@@ -0,0 +1,215 @@
+    interpolation and then a scaling to correct multiplicity
+    on shared ldofs. This encapsulates those two in one object
+    using the MFEM Operator interface. */
+class MFEMCeedInterpolation : public mfem::Operator
```
I named it `AlgebraicInterpolation`.

YohannDudouit commented on examples/ex1.cpp:
```diff
@@ -207,12 +211,20 @@ int main(int argc, char *argv[])
    {
       if (UsesTensorBasis(fespace))
       {
-         OperatorJacobiSmoother M(a, ess_tdof_list);
-         PCG(*A, M, B, X, 1, 400, 1e-12, 0.0);
+         if (algebraic_ceed)
```
Do we only want to use `algebraic_ceed` when we use tensor basis? The logic was different before.

barker29 replied:
Yes, it only works with tensor basis. The logic in `ex1` is now the same as in `ex1p` (which was correct before).

YohannDudouit commented on fem/ceed/operator.hpp:
```diff
@@ -33,9 +33,17 @@ class Operator : public mfem::Operator
    void Mult(const mfem::Vector &x, mfem::Vector &y) const override;
    void AddMult(const mfem::Vector &x, mfem::Vector &y) const;
    void GetDiagonal(mfem::Vector &diag) const;
+   mfem::Operator * SetupRAP(const mfem::Operator *Pi,
+                             const mfem::Operator *Po)
+   { return mfem::Operator::SetupRAP(Pi, Po); }
```
Do you really need to redefine this method? Isn't it already inherited? Or wouldn't `using mfem::Operator::SetupRAP;` suffice?

barker29 replied:
You're right, fixed.

YohannDudouit commented on fem/ceed/algebraic.cpp:
```diff
@@ -934,8 +934,6 @@ AlgebraicSolver::AlgebraicSolver(BilinearForm &form,
    MFEM_VERIFY(DeviceCanUseCeed(),
                "This object requires a Ceed device");
-   MFEM_VERIFY(form.GetAssemblyLevel() == AssemblyLevel::PARTIAL,
-               "AlgebraicSolver requires partial assembly");
```
```c++
MFEM_VERIFY(form.GetAssemblyLevel() == AssemblyLevel::PARTIAL || form.GetAssemblyLevel() == AssemblyLevel::NONE, "AlgebraicSolver requires partial assembly or fully matrix-free.");
MFEM_VERIFY(UsesTensorBasis(*fespaces), "AlgebraicSolver requires tensor product basis functions.");
```

YohannDudouit reviewed (approved):
The journey was long, but the result is really good I think, thank you for this contribution @barker29 . I don't know if this is expected but there is even less iterations in 3D than 2D.
```
./ex1 -pa -d ceed-cuda -o 8 -m ../data/fichera.mesh -a
Options used:
   --mesh ../data/fichera.mesh
   --order 8
   --no-static-condensation
   --partial-assembly
   --device ceed-cuda
   --algebraic
   --visualization
Device configuration: ceed-cuda,cuda,cpu
libCEED backend: /gpu/cuda/gen
Memory configuration: host-std,cuda
Number of finite element unknowns: 14877441
Size of linear system: 14877441
   Iteration :   0  (B r, r) = 0.0409363
   Iteration :   1  (B r, r) = 0.050941
   Iteration :   2  (B r, r) = 0.0183382
   Iteration :   3  (B r, r) = 0.00527929
   Iteration :   4  (B r, r) = 0.00187397
   Iteration :   5  (B r, r) = 0.000376886
   Iteration :   6  (B r, r) = 5.73609e-05
   Iteration :   7  (B r, r) = 2.51004e-05
   Iteration :   8  (B r, r) = 5.61375e-06
   Iteration :   9  (B r, r) = 8.78622e-07
   Iteration :  10  (B r, r) = 1.52965e-07
   Iteration :  11  (B r, r) = 1.28818e-08
   Iteration :  12  (B r, r) = 1.40207e-09
   Iteration :  13  (B r, r) = 2.83427e-10
   Iteration :  14  (B r, r) = 4.66597e-11
   Iteration :  15  (B r, r) = 9.40454e-12
   Iteration :  16  (B r, r) = 1.70229e-12
   Iteration :  17  (B r, r) = 1.69683e-13
   Iteration :  18  (B r, r) = 1.57934e-14
Average reduction factor = 0.45204
```

v-dobrev commented:
I see that `ex1`/`ex1p` is still modified in this PR. I think most reviewers expressed preference to not modify them and showcase the new capability elsewhere. From the options:
1. keep the changes in `ex1`/`ex1p` that illustrate the new solver
2. create a new example, e.g. `examples/ceed/{ex1,ex1p}.cpp`, to illustrate the new solver
3. add an option to `ex26`/`ex26p` to use the new solver,
4. other suggestions?

what is everyone's preference?

barker29 commented:
I prefer Option 1 (as I've said), it's only ~10 lines of changes. Option 2 would mean two very near duplicates of ex1/ex1p that we would need to maintain, so I don't like this. Option 3 is my second choice, but more or less the whole point of this PR is that you can do something like what's in `ex26` automatically without all the machinery introduced there. That new capability is more clearly demonstrated if it's shown in `ex1` rather than `ex26`.

YohannDudouit commented:
Option 1 is fine with me as the changes are minimal.

tzanio commented:
I have a slight preference for `4` but I'm ok with `1` too.

v-dobrev commented:
> I have a slight preference for `4` but I'm ok with `1` too.

Can you clarify what you mean by 4 (what is you "other suggestion")? Or was this a typo or a joke?

pazner commented:
I think this is a good case for resurrecting PR #942.

tzanio commented:
Sorry, I meant `3`... I also agree with Will about `Example 0`

v-dobrev commented:
My preference is for 2, since I don't like the idea of introducing dependency on an optional library in the first example.

tzanio commented:
> My preference is for 2, since I don't like the idea of introducing dependency on an optional library in the first example.

I agree, but I think our first example serves also the purpose of the "H1 Poisson" example and naturally is/will be becoming complex.

I think we should just add an example 0 as the simplest example. Lets discuss that in #942

barker29 commented:
I'm not sure what I'm supposed to do now, so I intend to bring this up in today's PR meeting. If you want to save time in that meeting you can just tell me what to do here and I will do it, even if it's something I don't like.

tzanio commented:
I don't think you should do anything yet, lets discuss today...

tzanio commented:
Re-merged in `next` for testing ...

tzanio commented:
LGTM

tzanio reviewed (approved):
Thank you so much for this important contribution @barker29 !

@YohannDudouit, @pazner, @dylan-copeland and @v-dobrev -- thank you for reviewing!
